homework/homework_8.ipynb

```diff
@@ -1,671 +1,827 @@
-    "        constraints=[S @ v == 0, v >= 0, v <= vmax_clipped],\n",
-    "    )\n",
-    "\n",
-    "    if problem.solve() < 0.2 * g_star:\n",
-    "        synthetic_lethals.append((i, j))\n",
-    "        \n",
-    "synthetic_lethals = np.array(synthetic_lethals)"
-   ]
-  },
-  {
-   "cell_type": "code",
-   "execution_count": 14,
-   "metadata": {},
-   "outputs": [],
-   "source": [
-    "# Add 1 since the indices in the question are 1-based not 0-based\n",
-    "essentials += 1\n",
-    "\n",
-    "synthetic_lethals += 1"
-   ]
-  },
-  {
-   "cell_type": "code",
-   "execution_count": 15,
-   "metadata": {},
-   "outputs": [
+  "cells": [
     {
-     "name": "stdout",
-     "output_type": "stream",
-     "text": [
-      "Essential genes:\n"
-     ]
+      "cell_type": "code",
+      "execution_count": 1,
+      "metadata": {
+        "id": "N8ioBZhNu4T5"
+      },
+      "outputs": [],
+      "source": [
+        "import itertools\n",
+        "\n",
+        "import pandas as pd\n",
+        "import numpy as np\n",
+        "import numpy as np\n",
+        "import cvxpy as cp\n",
+        "import matplotlib.pyplot as plt\n",
+        "\n",
+        "plt.style.use(\"seaborn\")"
+      ]
     },
     {
-     "data": {
-      "text/plain": [
-       "array([1, 9])"
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "ZfOU_2j4u4T_"
+      },
+      "source": [
+        "# Question 1: Gradient and Newton methods"
       ]
-     },
-     "metadata": {},
-     "output_type": "display_data"
     },
     {
-     "name": "stdout",
-     "output_type": "stream",
-     "text": [
-      "Synthetic lethals:\n"
-     ]
+      "cell_type": "code",
+      "execution_count": 2,
+      "metadata": {
+        "id": "LVCKEZwnu4UC"
+      },
+      "outputs": [],
+      "source": [
+        "n = 100\n",
+        "m = 200\n",
+        "np.random.seed(1)\n",
+        "A = np.random.normal(size=(m, n))"
+      ]
     },
     {
-     "data": {
-      "text/plain": [
-       "array([[2, 3],\n",
-       "       [2, 7],\n",
-       "       [4, 7],\n",
-       "       [5, 7]])"
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "GHmkmyx3u4UD"
+      },
+      "source": [
+        "Use cvxpy to get an accurate p_star which we'll use to plot error."
       ]
-     },
-     "metadata": {},
-     "output_type": "display_data"
-    }
-   ],
-   "source": [
-    "print(\"Essential genes:\")\n",
-    "display(essentials)\n",
-    "print(\"Synthetic lethals:\")\n",
-    "display(synthetic_lethals)"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "# Question 4: Online advertising displays"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "### Convert Matlab data generation to numpy\n",
-    "In this case, we're bound to get different answers because we're seeding the RNG, and we've translated it from Matlab,\n",
-    "so the RNG will be different."
```

Comment:
Can you remove this line, now that we're no longer generating random data?

Reply:
Done.